docs/proposals/loadbalancing.md

```diff
@@ -0,0 +1,338 @@
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# L7 Loadbalancing and ingress
+
+## Abstract
+
+This document proposes a basic L7 loadbalancing api for kubernetes 1.1. Expect hand-waving for anything not in scope.
+
+## Motivation
+
+The proposed api changes are largely motivated by the confusion surrounding kubernetes ingress and loadbalancing witnessed on [user forums](#appendix). Users tend to reason about traffic that originates outside the cluster at the application layer. The current kubernetes api lacks L7 support alltogether, and has a very opionated networking model. This leads to an impedance mismatch high enough to drive them to port brokering.
+
+## Use Cases
+
+It goes without saying that loadbalancers play a vital role in maintaining uptime and client connectivity. Kubernetes needs an easy, HA solution for routing traffic from an external network into the cluster using application layer intelligence. Master proxy is not an acceptable solution.
+
+## Scope
+
+This proposal is limited to the kubernetes 1.1 release. L7 loadbalancing is a vast topic and we cannot commit to an api without iterating on community feedback.
+
+__What's in scope for 1.1__
+* A single application using tls termination or raw http should not need more than 1 loadbalancer to dispatch multiple url endpoints to different backend services. Eg:
+
+  ```
+  www.example.com -> |terminate ssl| 178.91.123.132 -> / foo    s1
+                                                       / bar    s2
+                                                       / foobar s3
+  ```
+
+* A singleton service can continue to ask for L4 loadbalancing. Whether the request results in a new ip, or a new port on an existing ip, is left to the loadbalancer controller. The GCE controller for example, would give you a new ip.
+
+__What's not in scope for 1.1 (though it might be discussed below for context)__
+* Multiple hostnames per l7 loadbalancer.
+* Multiple certs per l7 loadbalancer.
+* Exposing loadbalancing algorithms/persistence options through the api.
+* A plugin model for multiple loadbalancers per cluster (i.e in 1.1, all paths will get claimed by the one loadbalancer controller the admin configures).
+* Users requesting loadbalancers (i.e a claims model).
+
+## Proposal
+
+The following proposal goes beyond the scope of 1.1 just so we don't make myopic decisions about the API. The tl;dr proposal for 1.1 is to define an IngressPoint resource type and write an L7 GCE load balancer controller that satisfies ingress requests. A draft of the L7 controller is available [here](https://github.com/kubernetes/kubernetes/pull/12825).
+
+__Terminology__:
+* **Ingress points**: A resource representing a collection of inbound connections from the external network that would be satisfied by a load balancer. Similar to GCE's `UrlMaps` or OpenShift's `Routes`.
+* **Claim**: A resource that represents a claim on an IP address.
```

Comment:
Could claims also be used for ingress point names ala more than 1 request for ```"claiming-to-own"```  ```www.example.org``` - it could still be that ```usain bolt``` wins (first one in) but the others may get rejected/alloted another [aka a generated] name based on the whatever policies are in place.